source/mir/math/stat.d

```diff
@@ -1626,6 +1627,704 @@ unittest
 }
 
+struct MapSummator(alias fun, T, Summation summation) 
+    if(isMutable!T)
+{
+    ///
+    Summator!(T, summation) mapSummator;
+
+    ///
+    F sum(F = T)() @property
+    {
+        return cast(F) mapSummator.sum;
+    }
+    
+    ///
+    void put(Range)(Range r)
+        if (isIterable!Range)
+    {
+        static if (hasShape!Range)
+        {
+            import mir.ndslice.topology: map;
+            mapSummator.put(r.map!fun);
+        }
+        else
+        {
+            foreach(x; r)
+            {
+                mapSummator.put(fun(x));
+            }
+        }
+    }
+
+    ///
+    void put()(T x)
+    {
+        mapSummator.put(fun(x));
+    }
+}
+
+///
+version(mir_test)
+@safe pure nothrow
+unittest
+{
+    import mir.ndslice.slice: sliced;
+    import mir.math.common: powi;
+
+    alias f = (double x) => (powi(x, 2));
+    MapSummator!(f, double, Summation.pairwise) x;
+    x.put([0.0, 1, 2, 3, 4].sliced);
+    assert(x.sum == 30.0);
+    x.put(5);
+    assert(x.sum == 55.0);
+}
+
+version(mir_test)
+@safe pure nothrow
+unittest
+{
+    import mir.ndslice.slice: sliced;
+
+    alias f = (double x) => (x + 1);
+    MapSummator!(f, double, Summation.pairwise) x;
+    x.put([0.0, 1, 2, 3, 4].sliced);
+    assert(x.sum == 15.0);
+    x.put(5);
+    assert(x.sum == 21.0);
+}
+
+version(mir_test)
+@safe pure @nogc nothrow
+unittest
+{
+    import mir.ndslice.slice: sliced;
+
+    alias f = (double x) => (x + 1);
+    MapSummator!(f, double, Summation.pairwise) x;
+    static immutable a = [0.0, 1, 2, 3, 4];
+    x.put(a.sliced);
+    assert(x.sum == 15.0);
+    x.put(5);
+    assert(x.sum == 21.0);
+}
+
+version(mir_test)
+@safe pure
+unittest
+{
+    import mir.ndslice.slice: sliced;
+    import mir.ndslice.fuse: fuse;
+
+    alias f = (double x) => (x + 1);
+    MapSummator!(f, double, Summation.pairwise) x;
+    auto a = [
+        [0.0, 1, 2],
+        [3.0, 4, 5]
+    ].fuse;
+    auto b = [6.0, 7, 8].sliced;
+    x.put(a);
+    assert(x.sum == 21.0);
+    x.put(b);
+    assert(x.sum == 45.0);
+}
+
+/++
+Variance algorithms.
+
+See Also:
+    $(WEB en.wikipedia.org/wiki/Algorithms_for_calculating_variance, Algorithms for calculating variance).
++/
+enum VarianceAlgo
+{
+    /++
+    Performs Welford's online algorithm for updating variance. When
+    calculating variance initially (not an update), the two-pass algorithm is 
+    used, whereby the input is first centered and then the sum of squares is 
+    calculated from that.    
+    +/
+    online,
+    
+    /++
+    Calculates variance using E(x^^2) - E(x)^2 (alowing for adjustments for 
+    population/sample variance). This algorithm can be numerically unstable.
+    +/
+    naive
+}
+
+private
+@safe pure nothrow @nogc
+T square(T)(scope const T x)
+{
+    import mir.internal.utility: isFloatingPoint;
+    import mir.math.common: powi;
+
+    static if (isFloatingPoint!T) {
+        return powi(x, 2);
```

Comment:
What is the reason to use `powi` and the `square` itslef?
Why not use just `MapSummator!(naryFun!"a * a", T, summation)`?

Reply:
I assumed that because `powi` used llvm intrinsics that it would be faster. Happy to change.